## Importing Libraries 

In [8]:
import pandas as pd
import numpy as np
import os 

## Exercise 4.5 - Practice Data Cleaning

In [9]:
df_ords = pd.read_csv(r"C:\Users\Windows\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv", index_col = 0)

In [10]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       hour_of_day_ordered  days_since_prior_order  
count         3.421083e+06            3.214874e+06  
mean          1.345202e+01            1.111484e+01  
std           4.226088e+00            9.206737e+00  
min           0.000000e+00            0.000000e+00  
25%           1.000000e+01            4.000000e+00  
50%           1.300000e+01            7.000000e+00  
75%           1.600000e+01            1.500000e+01  
max           2.300000e+01            3.000000e+01

In [11]:
# create a dataframe - to create a custom frame to serach your dataframe for mixed-type columns
df_test = pd.DataFrame()

In [12]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# check for mixed types 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [15]:
df_test['mix'] = df_test['mix'].astype('str')

In [16]:
df_test['mix'].dtype

dtype('O')

In [17]:
#Find Missing Values
df_prods = pd.read_csv(r"C:\Users\Windows\Instacart Basket Analysis\02 Data\Original Data\products.csv")

In [18]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [19]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
df_nan= df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
# Create a new dataframe without null values - all null values were in product_name
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
df_prods_clean.isnull()

product_id  product_name  aisle_id  department_id  prices
0           False         False     False          False   False
1           False         False     False          False   False
2           False         False     False          False   False
3           False         False     False          False   False
4           False         False     False          False   False
...           ...           ...       ...            ...     ...
49688       False         False     False          False   False
49689       False         False     False          False   False
49690       False         False     False          False   False
49691       False         False     False          False   False
49692       False         False     False          False   False

[49677 rows x 5 columns]

In [24]:
# Create dataframe with full duplciates
df_prods_clean.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Length: 49677, dtype: bool

In [25]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

In [30]:
df_prods_clean.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

## Task 4.5 Begins

In [31]:
# step 1 - consistentcy check on df_prods dataframe - see above

In [32]:
# step 2 -  Run the df.describe() function on your df_ords dataframe.

df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       hour_of_day_ordered  days_since_prior_order  
count         3.421083e+06            3.214874e+06  
mean          1.345202e+01            1.111484e+01  
std           4.226088e+00            9.206737e+00  
min           0.000000e+00            0.000000e+00  
25%           1.000000e+01            4.000000e+00  
50%           1.300000e+01            7.000000e+00  
75%           1.600000e+01            1.500000e+01  
max           2.300000e+01            3.000000e+01

##### there is a wide range in the order hour of the day. The maximum hour is at 11pm.  The average order hour is around  13. 45 hr or around 1:30pm. There is a large difference between maximum and the mean.

In [33]:
# step 3 check for mixed types 
# please see the process broken down into steps:
df_ords.columns.tolist()

['order_id',
 'user_id',
 'order_number',
 'orders_day_of_week',
 'hour_of_day_ordered',
 'days_since_prior_order']

In [34]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### No mixed-type data detected. 

In [35]:
# step 4  If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### No mixed-data type detected. Not applicable. 

In [36]:
# step 5  Run a check for missing values in your df_ords dataframe.

df_ords.isnull().sum()


order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_day_ordered            0
days_since_prior_order    206209
dtype: int64

In [37]:
# step 6 Address the missing values using an appropriate method

# find median
df_ords.median()

order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
hour_of_day_ordered            13.0
days_since_prior_order          7.0
dtype: float64

In [38]:
df_ords['days_since_prior_order'].fillna(7,inplace=True) 

In [39]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
hour_of_day_ordered       0
days_since_prior_order    0
dtype: int64

#### I chose imputation with the median value. The reason is that 206,209 missing values out of 3,421,083 total values (which is the count of total user IDs) is 6% of total values. As we learned in the earlier lessons, 5% is the cut-off for removing data. Therefore, I decided to keep these 200,000+ rows and imputed the missing valeus with the median value (7). I  then rechecked the tally of missing values, and the missing values were removed successfully from "days_since_prior_order". 


In [40]:
# step 7 Run a check for duplicate values in your df_ords data (look for full duplicates)

df_ords.duplicated()

0          False
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

In [41]:
# create a dataframe of duplicates

df_ords_dups = df_ords[df_ords.duplicated()]

In [42]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, hour_of_day_ordered, days_since_prior_order]
Index: []

#### No duplicates found 

In [43]:
# step 8 Address the duplicates using an appropriate method

#### No duplicates found. Not applicable.

In [44]:
path = r"C:\Users\Windows\Instacart Basket Analysis"

In [45]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords_cleaned.csv'))

In [46]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods_cleaned.csv'))